File: tests/XRoadLib.Tests/XRoadSerializableTest.fs
namespace XRoadLib.Tests

open FsUnit
open NUnit.Framework
open XRoadLib.Serialization
open XRoadLib.Tests.Contract

[<TestFixture>]
module XRoadSerializableTest =
    let [<Literal>] memberName = "Member"
    let [<Literal>] noMemberName = "No Member"

    [<Test>]
    let ``adds member names after serialization`` () =
        let instance = Class1()

        instance.IsSpecified(memberName) |> should be False
        instance.IsAcceptedByTemplate(memberName) |> should be False
        instance.IsSpecified(noMemberName) |> should be False
        instance.IsAcceptedByTemplate(noMemberName) |> should be False

        let serializable = instance :> IXRoadSerializable
        serializable.OnMemberDeserialized(memberName)

        instance.IsSpecified(memberName) |> should be True
        instance.IsAcceptedByTemplate(memberName) |> should be True
        instance.IsSpecified(noMemberName) |> should be False
        instance.IsAcceptedByTemplate(noMemberName) |> should be False

        serializable.SetTemplateMembers([ memberName ])

        instance.IsSpecified(memberName) |> should be True
        instance.IsAcceptedByTemplate(memberName) |> should be True
        instance.IsSpecified(noMemberName) |> should be False
        instance.IsAcceptedByTemplate(noMemberName) |> should be False


    [<Test>]
    let ``only in template when not serialized`` () =
        let instance = Class1()

        instance.IsSpecified(memberName) |> should be False
        instance.IsAcceptedByTemplate(memberName) |> should be False

        let serializable = instance :> IXRoadSerializable
        serializable.SetTemplateMembers([ memberName ])

        instance.IsSpecified(memberName) |> should be False
        instance.IsAcceptedByTemplate(memberName) |> should be True

        serializable.OnMemberDeserialized(memberName)

        instance.IsSpecified(memberName) |> should be True
        instance.IsAcceptedByTemplate(memberName) |> should be True

    [<Test>]
    let ``can handle member added multiple times`` () =
        let instance = Class1()

        instance.IsSpecified(memberName) |> should be False
        instance.IsAcceptedByTemplate(memberName) |> should be False

        let serializable = instance :> IXRoadSerializable
        serializable.SetTemplateMembers([ memberName; "muu"; "kolmas"; memberName ])

        instance.IsSpecified(memberName) |> should be False
        instance.IsAcceptedByTemplate(memberName) |> should be True

File: test/XRoadLib.Tests/Serialization/XRoadSerializableTest.cs
using XRoadLib.Serialization;
using XRoadLib.Tests.Contract;

namespace XRoadLib.Tests.Serialization;

public class XRoadSerializableTest
{
    private const string MemberName = "Member";
    private const string NoMemberName = "No Member";

    [Fact]
    public void AddsMemberNamesAfterSerialization()
    {
        var instance = new Class1();

        Assert.False(instance.IsSpecified(MemberName));
        Assert.False(instance.IsAcceptedByTemplate(MemberName));
        Assert.False(instance.IsSpecified(NoMemberName));
        Assert.False(instance.IsAcceptedByTemplate(NoMemberName));

        ((IXRoadSerializable)instance).OnMemberDeserialized(MemberName);

        Assert.True(instance.IsSpecified(MemberName));
        Assert.True(instance.IsAcceptedByTemplate(MemberName));
        Assert.False(instance.IsSpecified(NoMemberName));
        Assert.False(instance.IsAcceptedByTemplate(NoMemberName));

        ((IXRoadSerializable)instance).SetTemplateMembers(new[] { MemberName });

        Assert.True(instance.IsSpecified(MemberName));
        Assert.True(instance.IsAcceptedByTemplate(MemberName));
        Assert.False(instance.IsSpecified(NoMemberName));
        Assert.False(instance.IsAcceptedByTemplate(NoMemberName));
    }

    [Fact]
    public void ExistsOnlyInTemplateWhenNotSerialized()
    {
        var instance = new Class1();

        Assert.False(instance.IsSpecified(MemberName));
        Assert.False(instance.IsAcceptedByTemplate(MemberName));

        ((IXRoadSerializable)instance).SetTemplateMembers(new [] { MemberName });

        Assert.False(instance.IsSpecified(MemberName));
        Assert.True(instance.IsAcceptedByTemplate(MemberName));

        ((IXRoadSerializable)instance).OnMemberDeserialized(MemberName);

        Assert.True(instance.IsSpecified(MemberName));
        Assert.True(instance.IsAcceptedByTemplate(MemberName));
    }

    [Fact]
    public void CanHandleMemberAddedMultipleTimes()
    {
        var instance = new Class1();

        Assert.False(instance.IsSpecified(MemberName));
        Assert.False(instance.IsAcceptedByTemplate(MemberName));

        ((IXRoadSerializable)instance).SetTemplateMembers(new [] { MemberName, "muu", "kolmas", MemberName });

        Assert.False(instance.IsSpecified(MemberName));
        Assert.True(instance.IsAcceptedByTemplate(MemberName));
    }
}